pkg/cloudprovider/providers/photon/photon.go

```diff
@@ -0,0 +1,473 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package photon
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"github.com/golang/glog"
+	"github.com/vmware/photon-controller-go-sdk/photon"
+	"gopkg.in/gcfg.v1"
+	"io"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/cloudprovider"
+	k8stypes "k8s.io/kubernetes/pkg/types"
+	"log"
+	"os/exec"
+	"strings"
+)
+
+const (
+	ProviderName = "photon"
+)
+
+// Global variable pointing to photon client, can be assigned to mock client in tests
+var Esxclient *photon.Client
+var logger *log.Logger = nil
+var overrideIP bool = false
+
+// Photon is an implementation of cloud provider Interface for Photon Controller.
+type PCCloud struct {
+	cfg *PCConfig
+	// InstanceID of the server where this PCCloud object is instantiated.
+	localInstanceID string
+	// local $HOSTNAME
+	localHostname string
+	// hostname from K8S, could be overridden
+	localK8sHostname string
+	// Photon Project ID
+	projID string
+	//	instance      *Instances
+	cloudprovider.Zone
+}
+
+type PCConfig struct {
+	Global struct {
+		CloudTarget       string `gcfg:"target"`
+		IgnoreCertificate bool   `gcfg:"ignoreCertificate"`
+		Tenant            string `gcfg:"tenant"`
+		Project           string `gcfg:"project"`
+		OverrideIP        bool   `gcfg:"overrideIP"`
+	}
+}
+
+// Disks is interface for manipulation with PhotonController Persistent Disks.
+type Disks interface {
+	// AttachDisk attaches given disk to given node. Current node
+	// is used when nodeName is empty string.
+	AttachDisk(pdID string, nodeName k8stypes.NodeName) error
+
+	// DetachDisk detaches given disk to given node. Current node
+	// is used when nodeName is empty string.
+	DetachDisk(pdID string, nodeName k8stypes.NodeName) error
+
+	// DiskIsAttached checks if a disk is attached to the given node.
+	DiskIsAttached(pdID string, nodeName k8stypes.NodeName) (bool, error)
+
+	// CreateDisk creates a new PD with given properties.
+	CreateDisk(volumeOptions *VolumeOptions) (pdID string, err error)
+
+	// DeleteDisk deletes PD.
+	DeleteDisk(pdID string) error
+}
+
+// VolumeOptions specifies capacity, tags, name and flavorID for a volume.
+type VolumeOptions struct {
+	CapacityGB int
+	Tags       map[string]string
+	Name       string
+	Flavor     string
+}
+
+func readConfig(config io.Reader) (PCConfig, error) {
+	if config == nil {
+		err := fmt.Errorf("no Photon Cloud Provider config file given")
```

Comment:
nit: "no Photon Cloud Provider config file given" => "Photon Cloud Provider config file is missing. Please restart kubelet with --cloud-provider=photon --cloud-config=[path_to_config_file]" options.